### PR TITLE
[Cosmos DB] Introduce new api version 2021-04-15

### PR DESCRIPTION
**Pull request by nichatur**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changes in this PR
This PR introduces a new resources for Cosmos DB data plane RBAC (SqlRoleDefinition and SqlRoleAssignment), and this is the only change compared to the old 2021-04-15.

PR for adding these in a previous preview API version: [Cosmos DB - Adding APIs for Role Definition and Role Assignment](https://github.com/Azure/azure-rest-api-specs/pull/10493)

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet. -- Service changes are already in Prod, awaiting Swagger rollout
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet. -- Mid April
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by filizt** [comment]
>     "CloudError": {

Is there a reference to this CloudError definition?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/cosmos-db.json:5233 in fa5f3f9. [](commit_id = fa5f3f958b720cd1cca0eeb2e5e5a8bff6f1148e, deletion_comment = False)

**Comment by filizt** [comment]
The new APIs in rbac.json looks good. Please check out linter errors mostly coming from included referenced swaggers. There are errors for not supporting "SystemData" property. Here's RPC document about it: https://github.com/Azure/azure-resource-manager-rpc/blob/2e1f2c11ce4cfdf309f9cc8ae71575cb8fb83319/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources

**Comment by filizt** [comment]
one more question - @nichatur have you discussed the RBAC additions with Identity team?

**Review by JeffreyRichter — APPROVED** [review]

**Comment by nichatur** [comment]
> one more question - @nichatur have you discussed the RBAC additions with Identity team?

Yep we have, the resources exposed here are already in Public Preview

**Comment by nichatur** [comment]
> > ```
> > "CloudError": {
> > ```
> 
> Is there a reference to this CloudError definition?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/cosmos-db.json:5233 in [fa5f3f9](https://github.com/Azure/azure-rest-api-specs/commit/fa5f3f958b720cd1cca0eeb2e5e5a8bff6f1148e). [](commit_id = [fa5f3f9](https://github.com/Azure/azure-rest-api-specs/commit/fa5f3f958b720cd1cca0eeb2e5e5a8bff6f1148e), deletion_comment = False)

We are following the same pattern as  [Microsoft.Storage](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/storage/resource-manager/Microsoft.Storage/stable/2021-02-01/file.json
)

This error model allows RPs to transition from the legacy error format to the new one without breaking the swagger spec

**Comment by nichatur** [comment]
> The new APIs in rbac.json looks good. Please check out linter errors mostly coming from included referenced swaggers. There are errors for not supporting "SystemData" property. Here's RPC document about it: https://github.com/Azure/azure-resource-manager-rpc/blob/2e1f2c11ce4cfdf309f9cc8ae71575cb8fb83319/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources

There will be RP wide changes that will introduce systemData for all Cosmos DB resources, since pretty much all of them lack this as of now

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "cosmos-db.json#/definitions/CloudError"
+            }
+          }
+        }
+      },
+      "put": {
```
Comment (j5lim) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json:78]: >"put": [](http://example.com/codeflow?start=6&length=6)

PUT always replaces the entire resource. Does this api handel a partial update of the resource? If so, it must be done via PATCH.
[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc009-use-put-for-update-of-entire-resource-ie-replace-use-patch-only-for-partial-update-of-resource](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc009-use-put-for-update-of-entire-resource-ie-replace-use-patch-only-for-partial-update-of-resource)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
```
Comment (j5lim) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json:259]: > "put": { [](http://example.com/codeflow?start=5&length=9)

There must be PATCH if the api supports a partial update of a resource.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json (hunk 1 above)
Comment (nichatur) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json:78]: No we do not handle partial updates. Hence we do not support PATCH

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json (hunk 2 above)
Comment (nichatur) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2021-04-15/rbac.json:259]: We do not support partial update.

**Review by j5lim — APPROVED** [review]
Approved from ARM side.

**Review by weidongxu-microsoft — APPROVED** [review]
Please double check error reported in `Cross Version BreakingChange`

**Comment by nichatur** [comment]
> Please double check error reported in `Cross Version BreakingChange`

We are aware of those errors. The difference between preview and GA version is intentional. We have gotten a breaking change approval in lieu of that.

**Comment by weidongxu-microsoft** [comment]
@nichatur 

Good. Let me know when you need it merged.

**Pull request merged**